src/selectors/prevThought.ts

```diff
@@ -0,0 +1,124 @@
+  return lastVisibleDescendant(state, appendToPath(path, lastContext.parentId))
+}
+
+/**
+ * Gets the last visible descendant of the previous sibling.
+ * If the previous sibling is hidden, recursively tries the next previous sibling.
+ */
+const getLastVisibleDescendantOfPreviousSibling = (
+  state: State,
+  path: Path,
+  currentSibling: Path | null = null,
+): Path | null => {
+  const currentPath = currentSibling || path
+  const prevSiblingThought = prevSibling(state, currentPath)
+  if (!prevSiblingThought) return null
+  // Skip hidden thoughts only when showHiddenThoughts is false
+  const isHidden = !state.showHiddenThoughts && !isVisible(state, prevSiblingThought)
+  if (!isHidden) {
+    const prevSiblingPath = appendToPath(parentOf(path), prevSiblingThought.id)
+    return lastVisibleDescendant(state, prevSiblingPath)
+  }
+
+  const prevPath = appendToPath(parentOf(path), prevSiblingThought.id)
+  return getLastVisibleDescendantOfPreviousSibling(state, path, prevPath)
+}
+
+/** Recursively checks if any parent up to root is a context view root. */
+const findContextViewRoot = (state: State, path: Path): Path | null => {
+  if (isRoot(path)) return null
+  if (isContextViewActive(state, path)) return path
+  return findContextViewRoot(state, rootedParentOf(state, path))
+}
+
+/** Gets the previous thought in visual order. */
+const prevThought = (state: State, path?: Path): Path | null => {
+  if (!path) return null
+
+  const pathParent = rootedParentOf(state, path)
+  const isRootLevel = isRoot(pathParent)
+
+  // Check if we're in a context view
+  const inContextView = isContextViewActive(state, pathParent)
+  const hasActiveContextViews = Object.keys(state.contextViews).length > 0
+  // If in context view, try to get previous context first
+  if (inContextView) {
+    const prevContextThought = prevContext(state, path)
+    if (prevContextThought) return appendToPath(pathParent, prevContextThought.id)
+
+    // If no previous context, move to parent if not at root
+    return isRootLevel ? null : pathParent
+  }
+
+  const prevSiblingPath = getLastVisibleDescendantOfPreviousSibling(state, path)
+  // If we have active context views, check for context view boundaries
+  if (hasActiveContextViews && prevSiblingPath && path.length === 1) {
+    // Check if previous sibling is in a context view
+    const prevContextViewRoot = findContextViewRoot(state, prevSiblingPath)
+    if (prevContextViewRoot) return lastContext(state, prevContextViewRoot)
+  }
+
+  if (prevSiblingPath) return prevSiblingPath
+
+  // If at root level, stop here - don't traverse up to parent
+  if (isRootLevel) return null
+
+  // If not at root level and no previous sibling, check if parent is in context view
+  if (hasActiveContextViews && path.length === 1) {
+    const parentContextViewRoot = findContextViewRoot(state, pathParent)
+    if (parentContextViewRoot) return lastContext(state, parentContextViewRoot)
+  }
```

Comment:
This block is identical to one just a few lines above. There is probably a way to rearrange this code to eliminate the redundancy.